### ISSUE TITLE
[moved] Redefine `CARGO_TARGET_DIR` to be only an artifacts directory

### ISSUE THREAD

**Issue by kornelski** [rust-lang/rfcs#3664]
Moved https://github.com/rust-lang/cargo/issues/14125

**Comment by epage** [comment]
Is an RFC Issue the best place to move this conversation forward?  If it wasn't for the link here, I would never have seen this.

This has a lot of overlap with `--artfact-dir`, so I would lean towards that.

**Comment by poliorcetics** [comment]
Yes, the issue should be moved to cargo I think.

I'm not convinced at all this won't break backwards compatibility in some way.

> It makes ./target contain only workspace-unique files, which makes it justified for every workspace to have one.

And I don't want one in *any* cargo project while still keeping isolation, which is entirely different from what you are proposing.

> It enables moving registry deps to a shared build directory, without side effect of local projects overwriting each others' files. Sharing of dependencies matches users' expectation that the same dependencies shouldn't be redundantly rebuilt for each local project.

Once again, the RFC and original issue I wrote *does not* ask for that, it asks for the opposite: myself and many others want *separate* targets dirs for every project.

There are probably as many reasons as there are users for it but common ones are different sets of features amongst projects, sharing of build caches for specific projects, CI builds wanting to separate projects for security, or one project pinning 1.2.3 in a dep B of a dep A and the other project pinning 1.2.4: A can have the same version for both but it's dependencies won't and cargo is not made to handle the case at the moment.

You are fundamentally solving a different issue, one that the RFC I posted is *not* trying to solve.
